src/sql/src/func.rs

```diff
@@ -1263,28 +1452,78 @@ fn coerce_args_to_types(
-    ($p:ident...) => { ParamList::Variadic($p.into())};
+    ($p:ident...) => { ParamList::Variadic($p.into()) };
     ($($p:expr),*) => { ParamList::Exact(vec![$($p.into(),)*]) };
 }
 
+macro_rules! impl_def {
+    // Return type explicitly specified. This must be the case in situations
+    // such as:
+    // - Polymorphic functions: We have no way of understanding if the input
+    //   type affects the return type, so you must tell us what the return type
+    //   is.
+    // - Explicitly defined Operations whose returned expression does not
+    //   appropriately correlate to the function itself, e.g. returning a
+    //   UnaryFunc from a FuncImpl that takes two parameters.
+    // - Unimplemented/catalog-only functions
+    ($params:expr, $op:expr, $return_type:expr, $oid:expr) => {{
+        FuncImpl {
+            oid: $oid,
+            params: $params.into(),
+            op: $op.into(),
+            return_type: $return_type.into(),
+        }
+    }};
+    // Return type can be automatically determined as a function of the
+    // parameters.
+    ($params:expr, $op:expr, $oid:expr) => {{
+        let pcx = crate::plan::PlanContext::new(chrono::MIN_DATETIME, false);
+        let scx = StatementContext::new(None, &crate::catalog::DummyCatalog);
+        // This lifetime is compatible with more functions.
+        let qcx = QueryContext::root(&scx, QueryLifetime::OneShot(&pcx));
+        let ecx = ExprContext {
+            qcx: &qcx,
+            name: "dummy for builtin func return type eval",
+            scope: &Scope::empty(),
+            relation_type: &RelationType::empty(),
+            allow_aggregates: true,
+            allow_subqueries: false,
+            allow_windows: true,
+        };
+
+        let op = Operation::from($op);
+        let params = ParamList::from($params);
+        assert!(
+            !params.has_polymorphic(),
+            "loading builtin functions failed: polymorphic functions must have return types explicitly defined"
+        );
+
+        let cexprs = params.contrive_coercible_exprs();
+        let r = (op.0)(&ecx, cexprs, &params, vec![]).unwrap();
+        let return_type = r.return_type(&ecx, &params);
+
+        FuncImpl {
+            oid: $oid,
+            params,
+            op,
+            return_type,
+        }
+    }};
+}
+
 /// Constructs builtin function map.
 macro_rules! builtins {
     {
         $(
             $name:expr => $ty:ident {
-                $($params:expr => $op:expr, $oid:expr;)+
+                $($params:expr => $op:expr $(=> $return_type:expr)?, $oid:expr;)+
             }
         ),+
     } => {{
+
         let mut builtins = HashMap::new();
         $(
-            let impls = vec![
-                $(FuncImpl {
-                    oid: $oid,
-                    params: $params.into(),
-                    op: $op.into(),
-                },)+
-            ];
+            let impls = vec![$(impl_def!($params, $op $(,$return_type)?, $oid)),+];
```

Comment:
Ahhhh--this was what I couldn't get working that I bailed out and used token trees. tysm for this!